cpp/include/raft/neighbors/detail/nn_descent_batch.cuh

```diff
@@ -0,0 +1,684 @@
+
+      raft::neighbors::brute_force::fused_l2_knn<T, IdxT, raft::row_major, raft::row_major>(
+        res,
+        raft::make_const_mdspan(centroids),
+        raft::make_const_mdspan(d_dataset_batch.view()),
+        nearest_clusters_idx.view(),
+        nearest_clusters_dist.view(),
+        metric);
+      raft::copy(global_nearest_cluster.data_handle() + i * batch_size * k,
+                 nearest_clusters_idx.data_handle(),
+                 batch_size_ * k,
+                 resource::get_cuda_stream(res));
+    }
+  } else {  // data on device
+    auto nearest_clusters_idx =
+      raft::make_device_matrix<IdxT, IdxT, raft::row_major>(res, num_rows, k);
+    auto nearest_clusters_dist =
+      raft::make_device_matrix<T, IdxT, raft::row_major>(res, num_rows, k);
+
+    raft::neighbors::brute_force::fused_l2_knn<T, IdxT, raft::row_major, raft::row_major>(
+      res,
+      raft::make_const_mdspan(centroids),
+      raft::make_device_matrix_view<const T, IdxT, row_major>(dataset, num_rows, num_cols),
+      nearest_clusters_idx.view(),
+      nearest_clusters_dist.view(),
+      metric);
+
+    raft::copy(global_nearest_cluster.data_handle(),
+               nearest_clusters_idx.data_handle(),
+               num_rows * k,
+               resource::get_cuda_stream(res));
+  }
+}
+
+//
+// global_nearest_cluster [num_rows X k=2] : top 2 closest clusters for each data point
+// inverted_indices [num_rows x k vector] : sparse vector for data indices for each cluster
+// cluster_size [n_cluster] : cluster size for each cluster
+// offset [n_cluster] : offset in inverted_indices for each cluster
+// Loads the data in batches onto device if data is on host for memory efficiency
+//
+template <typename IdxT = uint32_t>
+void get_inverted_indices(raft::resources const& res,
+                          size_t n_clusters,
+                          size_t& max_cluster_size,
+                          size_t& min_cluster_size,
+                          raft::host_matrix_view<IdxT, IdxT> global_nearest_cluster,
+                          raft::host_vector_view<IdxT, IdxT> inverted_indices,
+                          raft::host_vector_view<IdxT, IdxT> cluster_size,
+                          raft::host_vector_view<IdxT, IdxT> offset)
+{
+  // build sparse inverted indices and get number of data points for each cluster
+  size_t num_rows = global_nearest_cluster.extent(0);
+  size_t k        = global_nearest_cluster.extent(1);
+
+  auto local_offset = raft::make_host_vector<IdxT>(n_clusters);
+
+  max_cluster_size = 0;
+  min_cluster_size = std::numeric_limits<size_t>::max();
+
+  for (size_t i = 0; i < n_clusters; i++) {
+    cluster_size.data_handle()[i] = 0;
+    local_offset.data_handle()[i] = 0;
+  }
+
+  for (size_t i = 0; i < num_rows; i++) {
+    for (size_t j = 0; j < k; j++) {
+      IdxT cluster_id = global_nearest_cluster.data_handle()[i * k + j];
+      cluster_size.data_handle()[cluster_id]++;
+    }
+  }
```

Comment:
This section may benefit from an omp parallel region. `cluster_size(cluster_id)++` will have to be an atomic though. Overall, I don't know how beneficial this speedup will be so you can just leave this idea as a comment in the code so we can try later on if necessary